### PR TITLE
Faster calculation of geometric relations

### PR DESCRIPTION
**Pull request by patrickbr**
This PR summarizes our efforts of the last 6 months to significantly speed up the calculation and improve the quality of the geometric relations (contains and intersects). The most significant changes, given in approximate chronological order and without guarantee of completeness, are:

1. No geometries are stored in the R-Tree anymore, only reference IDs. This drastically reduces the R-Tree overhead.
2. Simplified outer and inner geometries of areas are used to decide contains and intersect relations
3. The contains-DAG is now build using approximate covered_by, based on area coverage. This greatly improves the contains-DAG quality in situations where floating point precision issues and/or precision issues in OSM prevent a contains of obviously contained geometries (like Baden-Württemberg in Germany).
4. Multipolygons are stored separately in the geometric index. Results retrieved from the geometric index are made unique before further processing. This avoids huge bounding boxes for scattered multipolygons (prominent example: Overseas France)
5. Inner/Outer simplification parameters for polygons are based linearly on the radius of a hypothetical circle with the area of the input polygon. This avoids extremely high simplification parameters for polygons with highly convoluted rings (e.g. complicated coastlines).
6. Each area is given a list of integer IDs corresponding to global grid cells ("box ids") that are fully contained in the area (of the ID is positive) or only intersected by the area (if the ID is negative). These IDs are also computed for way geometries, and on-the-fly for node geometries. Contains / intersect relations can then often be decided by intersecting the box id lists, without further geometric calculations.
7. For polygons with many anchor points, in addition to the box ID of 6, a "cutout" geometry (the geometric intersection between the grid cell and the input geometry) is also stored. If the box IDs intersect then reveals, for example, that way A lies completely in box id 453, but box is 453 is only partially contained in area B, it is enough to check A against the cutout of B for box id 453.
8. Semantic information from OSM is now used to skip some contains/intersect calculations. Example: if area A is an inner ring of area B, and we know that way C is in A, we can be sure that C is neither contained in B, nor intersects B.

In addition, the triple writing has been slightly sped up by the following changes:

1. For certain triples we have complete control over, we now use unsafe writing methods which skip any character checking and/or escaping.
2. The simplification of area geometries and the calculation of their area during the fact dump is now non-blocking.

With these changes, the full planet.osm with geometric relation triples can now be converted in about 8 hours (down from several months).

### CODE REVIEW

File: CMakeLists.txt
Comment (lehmann-4178656ch) [inline CMakeLists.txt:44]: I think the order of comments and commands got mangled

File: include/osm2rdf/config/Config.h
Comment (lehmann-4178656ch) [inline include/osm2rdf/config/Config.h:48]: Why 1/(π*20)?

File: src/config/Config.cpp
Comment (lehmann-4178656ch) [inline src/config/Config.cpp:163]: Indentation got mangled

File: src/osm/FactHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/FactHandler.cpp:63]: `BASE10` exists as `osm2rdf::osm::constants::BASE10_BASE`

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:259]: `ioDouglasPeucker` -> `innerOuterDouglasPeucker`? If so it would be a few characters longer but more readable as `io` could be input/output as well.

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:497]: This comment does not correspond to any line. Nearest would be 518 which replaces the index with a new one.

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:591]: What's 1? Could this be a bool or an enum?

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:1142]: As before, what's `areaFromType == 0`? A comment or enum could improve readability here.

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:1337]: `size() == 0` -> `empty()`?

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:1441]: Replace with `empty()`?

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:1552]: `empty()`?

**Review by lehmann-4178656ch — COMMENTED** [review]
Thanks for the work, I have some small remarks for now. I'll take a deeper look later.

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:87]: This was changed in in Way.h and should be changeable in Relation.h too. https://github.com/ad-freiburg/osm2rdf/blob/095ef8e6f0d9e46490232d7b4b5c5d9bc7117b73/include/osm2rdf/osm/Way.h#L50

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:139]: Just a comment for the future:
For oriented-bounding-boxes the convex hull is required if [rotating calipers](https://en.wikipedia.org/wiki/Rotating_calipers) is used. This could mean that everything but Nodes will need this geometry later.

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:150]: Maybe a small comment that this adds the envelope of every part of the a multipolygon with positive area. It's clear after looking at it, but it's not obious here why the distinction between `area.geom().size() >1` and `area.geom().size() <= 1` is needed. Is recalculating the Envelope that expensive?

Speaking of the `area.goem().size() <= 1` case: why is the evnelope added a second time? At the time of writing this seams to be the case that the first entry is always the bounding box of everything in the mulitpolygon and every subsequent entry for each part.

Is this done to simplify the code of later steps?

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:203]: Just to be sure: if a way has at least `2 * CHUNKSIZE` nodes it receives an additional bounding boxes for every parts of size `CHUNKSIZE` (except the last) which is later added to the r-tree? A small comment to answer why this is done would improve readability here by a huge point - this is related to the remark about `GeometryHandler<W>::area` where a short explanation why could improve the understanding too

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:327]: `enum` for MODE?

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/GeometryHandler.cpp:426]: `enum` for MODE in innerOuterDougasPeucker could improve readability here.

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [reply src/osm/GeometryHandler.cpp:505]: 518 moved to 526 but this is still relevant as the rtree is never emptied but replaced during `_spatialIndex = SpatialIndex(values.begin(), values.end());`

File: src/osm/OsmiumHandler.cpp
Comment (lehmann-4178656ch) [inline src/osm/OsmiumHandler.cpp:209]: Maybe we should task this away if boost is available as done for `osmArea.finalize();`?

File: src/osm/GeometryHandler.cpp
Comment (lehmann-4178656ch) [reply src/osm/GeometryHandler.cpp:87]: I think this is already changed in Relation.h -> maybe this comment could be removed?

File: src/osm/OsmiumHandler.cpp
Comment (patrickbr) [reply src/osm/OsmiumHandler.cpp:209]: I agree, will change this :)

File: src/osm/GeometryHandler.cpp
Comment (patrickbr) [reply src/osm/GeometryHandler.cpp:426]: I agree,

File: src/osm/GeometryHandler.cpp
Comment (patrickbr) [reply src/osm/GeometryHandler.cpp:203]: No, it is not added to the R-Tree (ways are never added to the R-Tree). These envelopes are used for *querying* the R-Tree in `indexQryIntersect()`. That is, instead of sending a huge intersect query for the entire bounding box of the way, we send several smaller queries to the R-Tree, aggregate the result and make it unique later on.

Comment (lehmann-4178656ch) [reply src/osm/GeometryHandler.cpp:203]: Makes sense, I quess it was a bit late when I looked at it. But I think adding a small explanation to the code would still be good.

File: src/osm/GeometryHandler.cpp
Comment (patrickbr) [reply src/osm/GeometryHandler.cpp:150]: I agree, I added a comment to clarify that. In general, calculating the bounding box is not expensive, but it felt wrong to iterate over thousands of anchor points just to generate a bounding box we already have :)

Yes, bounding boxes are stored two times for simple (non-multi) polygons to make later code easier.

File: src/osm/GeometryHandler.cpp
Comment (patrickbr) [reply src/osm/GeometryHandler.cpp:87]: Indeed!

File: src/osm/GeometryHandler.cpp
Comment (patrickbr) [reply src/osm/GeometryHandler.cpp:203]: I agree, I have already added the comment, will commit in a few minutes :D

**Comment by patrickbr** [comment]
@lehmann-4178656ch, I added some changes addressing your comments.

Another thing: building `planet.osm` on `ob` with the relation geometries took around 8.5 hours, so building these geometries doesn't seem to add much time.

However, there was a heavy increase in memory usage. I haven't measured this, but when I checked last evening (when the fact dump was at around 95%) it used 150 GB of virtual memory, and around 80 GB of actual memory. Is this because the required way geometries are all hold in memory, until the fact dump is finished?

**Comment by patrickbr** [comment]
(To further elaborate: we should address this once this branch has been merged into master)

**Review by lehmann-4178656ch — APPROVED** [review]
I think we can and should merge this now ... I can try to take a look later to find out why github does not offer the simple Squash & Merge button and claims that there are conflicts. If you have time you can merge this too.

**Comment by patrickbr** [comment]
Done, I will change the automatic generation for osm2rdf.cs.uni-freiburg.de to use the master branch from now on.

**Pull request merged**
